yjit/src/asm/mod.rs

```diff
@@ -0,0 +1,386 @@
+use std::collections::BTreeMap;
+use std::mem;
+
+// Lots of manual vertical alignment in there that rustfmt doesn't handle well.
+#[rustfmt::skip]
+pub mod x86_64;
+
+/// Pointer to a piece of machine code
+/// We may later change this to wrap an u32
+/// Note: there is no NULL constant for CodePtr. You should use Option<CodePtr> instead.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Debug)]
+#[repr(C)]
+pub struct CodePtr(*const u8);
+
+impl CodePtr {
+    pub fn raw_ptr(&self) -> *const u8 {
+        let CodePtr(ptr) = *self;
+        return ptr;
+    }
+
+    fn into_i64(&self) -> i64 {
+        let CodePtr(ptr) = self;
+        *ptr as i64
+    }
+
+    fn into_usize(&self) -> usize {
+        let CodePtr(ptr) = self;
+        *ptr as usize
+    }
+}
+
+impl From<*mut u8> for CodePtr {
+    fn from(value: *mut u8) -> Self {
+        assert!(value as usize != 0);
+        return CodePtr(value);
+    }
+}
+
+/// Compute an offset in bytes of a given struct field
+macro_rules! offset_of {
+    ($struct_type:ty, $field_name:tt) => {{
+        // Null pointer to our struct type
+        let foo = (0 as *const $struct_type);
+
+        unsafe {
+            let ptr_field = (&(*foo).$field_name as *const _ as usize);
```

Comment:
FYI this is a dereference of a null pointer, which is UB. It probably doesn't codegen to an actual read of address 0, so you don't get a fault, but it's UB nonetheless. If possible, it would be good to use the `memoffset` crate instead.

I don't know if this was copy-pasted from elsewhere, but `bindgen` has the same UB code pattern in it. I wouldn't be surprised if this code has spread all over.

Reply:
LLVM definitely doesn't read an address of 0 because this is an address computation, no memory is actually being read.

We've been trying to avoid using external crates which is the problem. I would like to use some sort of built in `offsetof` operator instead, it would look cleaner, but unfortunately it's only accessible through an external crate.

Reply:
While I greatly worry about pasting `unsafe` code from external crates (multiple times, I've fixed the same piece of UB in 2 or 3 crates because they all copy+pasted instead of having a dependency), `memoffset` is MIT and the source for `offset_of!` is right here: https://github.com/Gilnaa/memoffset/blob/66f4c956a72daabeafa87ba9aef339d9035d78b4/src/offset_of.rs#L90 and I think it would be better to copy+paste than definitely have UB.

Reply:
Thanks for letting us know! This macro is actually unused and I'm in the process of deleting it while addressing other dead code lints. We have `#[allow(unused_macros)]` at the moment which hides the lint. We'll be sure to try for a UB free implementation when we need one. :)